### PR TITLE
UnitedDeployment: revision maintenance

### PR DESCRIPTION
**Pull request by wu8685**

### CODE REVIEW

File: pkg/controller/uniteddeployment/revision.go [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+// assumes that revisions has been sorted by Revision.
+func nextRevision(revisions []*apps.ControllerRevision) int64 {
+	count := len(revisions)
+	if count <= 0 {
+		return 1
+	}
+	return revisions[count-1].Revision + 1
+}
+
+func getUnitedDeploymentPatch(ds *appsalphav1.UnitedDeployment) ([]byte, error) {
```
Comment (Fei-Guo) [inline pkg/controller/uniteddeployment/revision.go]: ds ->ud?

File: pkg/controller/uniteddeployment/revision.go [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+	}
+	var raw map[string]interface{}
+	err = json.Unmarshal(dsBytes, &raw)
+	if err != nil {
+		return nil, err
+	}
+	objCopy := make(map[string]interface{})
+	specCopy := make(map[string]interface{})
+
+	// Create a patch of the DaemonSet that replaces spec.template
```
Comment (Fei-Guo) [inline pkg/controller/uniteddeployment/revision.go]: Wrong comment?

File: pkg/controller/uniteddeployment/utils/ref_manager.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+	}
+
+	claimObjs := []metav1.Object{}
+	errlist := []error{}
+	for _, obj := range objs {
+		ok, err := mgr.claimObject(obj, match)
+		if err != nil {
+			errlist = append(errlist, err)
+			continue
+		}
```
Comment (Fei-Guo) [inline pkg/controller/uniteddeployment/utils/ref_manager.go]: remove `continue` and change to else if ok ?

File: pkg/controller/uniteddeployment/revision.go [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+		hf.Write([]byte(strconv.FormatInt(int64(*probe), 10)))
+	}
+	return rand.SafeEncodeString(fmt.Sprint(hf.Sum32()))
+}
+
+// newRevision creates a new ControllerRevision containing a patch that reapplies the target state of set.
+// The Revision of the returned ControllerRevision is set to revision. If the returned error is nil, the returned
+// ControllerRevision is valid. StatefulSet revisions are stored as patches that re-apply the current state of set
+// to a new StatefulSet using a strategic merge patch to replace the saved state of the new StatefulSet.
+func (r *ReconcileUnitedDeployment) newRevision(cd *appsalphav1.UnitedDeployment, revision int64, collisionCount *int32) (*apps.ControllerRevision, error) {
```
[the remark this answers is not in the document]
Comment (wu8685) [reply pkg/controller/uniteddeployment/revision.go]: Done

File: pkg/controller/uniteddeployment/revision.go (hunk 1 above)
Comment (wu8685) [reply pkg/controller/uniteddeployment/revision.go]: done

File: pkg/controller/uniteddeployment/revision.go (hunk 2 above)
Comment (wu8685) [reply pkg/controller/uniteddeployment/revision.go]: done

File: pkg/controller/uniteddeployment/utils/ref_manager.go (hunk 3 above)
Comment (wu8685) [reply pkg/controller/uniteddeployment/utils/ref_manager.go]: done

**Comment by FillZpp** [comment]
Take a look at  Travis CI `continuous-integration/travis-ci/pr`. 
It seems that `pkg/controller/uniteddeployment` unit-test failed.

**Pull request merged**